rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+
+## Declaration Engine
+
+The declaration engine takes the concept of constraints and unification and applies it to declarations.
+
+For example, when type checking a function application, the compiler does not need to care about the contents of the body of the function, instead it is sufficient to create a constraint referring to "the return type of the function".
+
+For example, given a typed function declaration `add`, we add this declaration to the declaration engine. Then when type checking a function application of `add`, instead of inlining the function body, we create constraints referring to "the body of `add`" and "the return type of `add`". When type checking a _function application_
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This design will take after the design of the [type engine](https://github.com/FuelLabs/sway/blob/2816c13698a35752136d4843dbfe5e1b95c26e10/sway-core/src/type_engine/engine.rs#L14). During type checking, the compiler will add information about declarations to the declaration engine. Then, in places where that declaration is used, a value representing that declaration can be generated in place of inlining the declaration.
+
+### Definition
+
+During type checking of a declaration, the compiler will add information about that declaration to the declaration engine:
+
+```rust
+struct DeclarationEngine {
+    // function name to typed function declaration
+    // this example code does not consider Path's, but we could use
+    // a BTreeMap with Path's as the key
+    functions: HashMap<String, TypedFunctionDeclaration>,
+
+    // struct name to typed struct declaration
+    // again, this does not consider Path's
+    structs: HashMap<String, TypedStructDeclaration>,
+
+    // enums and traits
+}
+
+impl DeclarationEngine {
+    fn add_function(&mut self, function: TypedFunctionDeclaration) -> DeclarationDefinition {
+        // add to self and return DeclarationDefinition
+    }
+
+    fn add_struct(&mut self, r#struct: TypedStructDeclaration) -> DeclarationDefinition {
+        // add to self and return DeclarationDefinition
+    },
+
+    // enums and traits
+}
+
+enum DeclarationDefinition {
+    // does not consider Path's
+    Function(String),
+
+    // does not consider Path's
+    Struct(String),
+
+    // enums and traits
+}
+```
+
+Then in the typed AST, that declarations could look like (in pseudocode):
+
+```rust
+contract;
+
+/// imports
+
+DeclarationDefinition::Function(<function name>)
+DeclarationDefinition::Struct(<struct name>)
+
+/// storage declaration
+
+DeclarationDefinition::Function("main")
+
+```
+
+Because inlining is still needed for the IR, these instances of `DeclarationDefinition` would be replaced in an additional step at the end of type checking called the [resolution step](https://github.com/FuelLabs/sway/issues/1820).
```

Comment:
Inlining isn't really needed for the backend and the resolution step you link to there is more about confirming that the AST is fully resolved/unified, no unknown types remain.

Is this a third definition of 'inlining'?  :slightly_smiling_face: 'Ensuring all generic types are replaced with concrete types'?

The back end just wants low level consistent fully-qualified naming for declarations and references.  So, for functions in a module it's easy -- `crate::module::fn_name` -- but for e.g., trait methods it's harder.  When we call `foo_bar.into()` we need a `crate::module::concrete_impl::into` name to call.